### PR TITLE
ENH Pyfolio integration

### PR DESCRIPTION
**Pull request by luca-s**
Issue #225 

Added performance.create_pyfolio_input, which create input suitable for pyfolio. Intended use:

```
factor_data = alphales.utils.get_clean_factor_and_forward_returns(factor, prices)

pf_returns = alphales.performance.create_pyfolio_input(factor_data,'1D',
                     long_short=True,
                     group_neutral=False,
                     quantiles=None,
                     groups=None)

pyfolio.tears.create_returns_tear_sheet(pf_returns)
```

Also, I  greatly improved the function that computes the cumulative returns as that is the base on which the pyfolio integration feature is built on.  Mainly I removed the legacy assumptions  which required the factor to be computed at specific frequency (daily, weekly etc) and also the periods had to be multiple of this frequency (2 days, 4 days etc). The function is now able to properly compute returns when there are gaps in the factor data or when we analyze an intraday factor.

### CODE REVIEW

**Comment by luca-s** [comment]
Oh, I forgot. To properly compute the portfolio returns I needed to know the length of the periods. Currently they are just integers and it is not known the units of those integers. So I changed the period column names from a simple integer to a pd.Timedelta string representation (1D, 5H, etc)

**Comment by luca-s** [comment]
I'll go through pep8 issues tomorrow.

**Comment by luca-s** [comment]
Ok, finished with pep8 issues. This PR is ready

File: alphalens/performance.py [hunk 1]
```diff
@@ -355,8 +496,10 @@ def quantile_turnover(quantile_factor, quantile, period=1):
         DataFrame with date, asset and factor quantile.
     quantile : int
         Quantile on which to perform turnover analysis.
-    period: int, optional
+    period: pandas.Timedelta or string
         Period over which to calculate the turnover
+        if 'period' is a string it must follow pandas.Timedelta constructor
+        format (.e.g '1 days', '1D', '30m', '3h', '1D1h', etc)
```
Comment (twiecki) [inline alphalens/performance.py]: `.e.g` -> `e.g.`

**Comment by twiecki** [comment]
This looks great!

**Comment by luca-s** [comment]
I am pretty sure there are still many issues because I haven't done many tests yet, especially because I cannot access Quantopian data on my local machine. Anyway, let's consider this an experimental feature for now and then we will improve it.

**Comment by luca-s** [comment]
I realized I wasn't clear in my previous comment. What I mean is that the PR is ready to be merged but I expect some issues with pyfolio as I haven't done extensive testing. Anyway I don't expect regressions with existing functionalities and I actually expect better compatibility with non-daily factors and event studies.

**Comment by luca-s** [comment]
@twiecki I am going to merge this if you agree

File: alphalens/utils.py [hunk 2]
```diff
@@ -680,8 +640,36 @@ def std_conversion(period_std):
...
 def get_forward_returns_columns(columns):
-    return columns[columns.astype('str').str.isdigit()]
+    pattern = re.compile(r"^(\d+([Dhms]|ms|us|ns))+$", re.IGNORECASE)
+    valid_columns = [(pattern.match(col) is not None) for col in columns]
+    return columns[valid_columns]
+
+
+def timedelta_to_string(timedelta):
+    """
+    Utility that converts a pandas.Timedelta to a string rappresentation
```
Comment (twiecki) [inline alphalens/utils.py]: representation

File: alphalens/utils.py [hunk 3]
```diff
@@ -680,8 +640,36 @@ def std_conversion(period_std):
...
 
 
 def get_forward_returns_columns(columns):
-    return columns[columns.astype('str').str.isdigit()]
+    pattern = re.compile(r"^(\d+([Dhms]|ms|us|ns))+$", re.IGNORECASE)
+    valid_columns = [(pattern.match(col) is not None) for col in columns]
+    return columns[valid_columns]
+
+
+def timedelta_to_string(timedelta):
```
Comment (twiecki) [inline alphalens/utils.py]: Does that really not exist inside pandas?

Comment (luca-s) [reply alphalens/utils.py]: Unfortunately no, the pandas version gives back the verbose version. e.g. '1 days 00:00:00'

File: alphalens/utils.py (hunk 2 above)
Comment (luca-s) [reply alphalens/utils.py]: thanks :)

**Comment by twiecki** [comment]
Does this change existing behavior?

**Comment by luca-s** [comment]
No, all scenarios that were working before are working now + the current behaviour does a better job in handling factors with gaps (when values are not computed every day, e.g. event study) and intraday factors.

The only user level visible change is the period column names that are now named "1D", "3D" (1 day, 3 days or "1H", 3H"  for 1 hour, 2 hours, etc. ) instead of "1", "2".

**Comment by twiecki** [comment]
OK, great. And users can now also have arbitrary periods, like hours.

I haven't reviewed the code carefully but I'm fine merging this as an experimental feature.

**Comment by twiecki** [comment]
Actually, let's see if @richafrank has someone who could do a code review.

**Comment by luca-s** [comment]
Actually arbitrary periods are possible even without this PR, for example a daily factor with 1 hours period or a weekly factor (computed on Monday for example) with a period of 3 days. With this kind of factors the results are computed correctly even though the turnover analysing is not so interesting, but the cumulative returns are wrong and this prevent the pyfolio analysis, this is why I fixed it and as I was already fixing the code I made sure it works with factors not  computed 
at a fixed frequency, e.g. event study

**Comment by abhijeetkalyan** [comment]
@luca-s @twiecki this looks neat! I had a couple questions around the UX:
- The name confused me a bit. When I saw `create_pyfolio_input`, I expected it to have positions and transactions as well. Since the function's output is the return stream, should we change its name to clarify that it's providing only returns, and not all of the other inputs that pyfolio consumes?
- If this is still an experimental feature as @twiecki mentioned in https://github.com/quantopian/alphalens/pull/227#issuecomment-351047163, should we add some kind of warning to inform users of that fact?

**Comment by luca-s** [comment]
You are right @abhijeetkalyan , the name is confusing a little. The reason is that I initially thought to return  `positions` and `transactions` as well but I haven't managed to do that yet. I am actively working on this so my plan is to improve `create_pyfolio_input` to make it returns `positions` and `transactions`  in another PR.  I hope to push that PR soon together with an example NB. If I manage to do that before we create a new tag I would avoid the warning, otherwise if we create a tag and release a new version of Alphalens before I finish working on this feature I will add a warning message. Sounds like a plan?

**Comment by luca-s** [comment]
I added an example NB and also I made `create_pyfolio_input` returns the benchmark returns too (as the factor universe mean returns).

**Comment by luca-s** [comment]
I also added a warning in `create_pyfolio_input` docstring:

>  WARNING: this API is still in experimental phase and input/output
>                      paramenters might change in the future 

I believe this PR is now ready to be merged

File: alphalens/performance.py [hunk 4]
```diff
@@ -68,7 +68,7 @@ def src_ic(group):
         grouper.append('group')
 
     ic = factor_data.groupby(grouper).apply(src_ic)
-    ic.columns = pd.Int64Index(ic.columns)
+    ic.columns = pd.Index(ic.columns)
```
[the remark this answers is not in the document]
Comment (luca-s) [reply alphalens/performance.py]: I guess you are right :)

File: alphalens/utils.py [hunk 5]
```diff
@@ -604,66 +613,17 @@ def common_start_returns(factor,
...
     """
-    return period_ret.add(1).pow(1. / period_ret.name).sub(1)
+    period_len = period_ret.name
+    conversion_factor = (pd.Timedelta(one_period_len) /
+                         pd.Timedelta(period_len))
+    return period_ret.add(1).pow(conversion_factor).sub(1)
 
 
-def std_conversion(period_std):
+def std_conversion(period_std, one_period_len):
```
[the remark this answers is not in the document]
Comment (luca-s) [reply alphalens/utils.py]: you are right!

**Comment by luca-s** [comment]
@prsutherland thanks for reminding me about the walk through NB I completely forgot about that. Regarding the  default values on the `period` parameters I am not sure... If I find a sensible default I would restore them.
The movement from integers to Timedelta strings for period was needed to carry information that otherwise would have been lost but some thorough testing is now required because this change could lead to some hidden and non-obvious bugs. Sometimes I think get_clean_factor_and_forward_returns should return an object  where we can store all the information we want together with factor_data instead of trying to save the information we need inside the DataFrame column names...but for now let's keep it this way.

**Comment by luca-s** [comment]
Unless there are other comments I am going to merge this PR soon.  I only need to finish updating the tear_sheet_walk_through NB.

**Comment by twiecki** [comment]
Looks great, @luca-s. Thanks!

**Pull request merged**